[dynamo] Fix support for nn.Parameter constructor (part 1)

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121610
* #120965
* #121609
* __->__ #120163
* #121086

This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

jansel committed:
[dynamo] Fix support for nn.Parameter constructor

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

albanD commented on test/inductor/test_cpu_repro.py:
```diff
@@ -2878,6 +2879,35 @@ def forward(self, x):
+            def forward(self, x):
+                self.conv.weight = nn.Parameter(self.conv_weight)
+                self.conv.bias = nn.Parameter(self.conv_bias, requires_grad=False)
+                self.conv.eval()
```
This is not cool haha
It should be a no-op though so not sure it changes much? Since you don't test it explicitly, I would remove it.

Unless you meant to call this on the batchnorm where it will change its behavior!

jansel replied:
This is just copied from the repro provided in https://github.com/pytorch/pytorch/issues/99569

I agree it doesn't make much sense, but want to make sure we don't break on that user report.

albanD commented on test/inductor/test_cpu_repro.py:
```diff
@@ -2878,6 +2879,35 @@ def forward(self, x):
+        func = Model2().to("cpu")
+
+        with torch.no_grad():
+            func.train(False)
```
This makes the eval() call above a noop?

jansel replied:
Yes

albanD commented on torch/_dynamo/output_graph.py:
```diff
@@ -1575,6 +1582,13 @@ def add_graph_finalizer(
+        if node.op == "placeholder":
+            return node.meta["grapharg"].example
+        assert node.op == "get_attr"
+        return self.nn_modules[node.target]  # type: ignore[index]
```
Is "get_attr" a nn Module specific op?

jansel replied:
Yeah, dynamo only uses `get_attr` for nn modules (and tensors stored on nn modules).

albanD commented on torch/_dynamo/variables/torch.py:
```diff
@@ -759,3 +764,50 @@ def handle_ntuple(value):
+        cg.load_import_from("torch.nn", "Parameter")
+        cg(data.source)
+        cg(variables.ConstantVariable(requires_grad))
+        cg.call_function(2, True)
```
I can't find the definition of this function in torch/_dynamo/codegen.py where PyCodegen is defined. Where is it?

jansel replied:
https://github.com/pytorch/pytorch/blob/bcf35c6ae62bb6560befa3550e37a8283944e5f4/torch/_dynamo/codegen.py#L323-L324

albanD replied:
Ho it was added last week and is not in the folder I was checking. Thanks!

albanD reviewed (commented):
I'm not the best person to review the dynamo code so this is mostly questions haha

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

albanD reviewed (approved):
The change sounds good to me. You might want a second opinion from someone familiar with Dynamo though.

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Fix support for nn.Parameter constructor (part 1)"


This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel commented:
@pytorchbot merge
